File: list/3-环形链表 II/solve2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# https://leetcode-cn.com/problems/linked-list-cycle-ii/
class Solution:
    # 这尼玛主要考数学
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head:
            return
        fast = head
        slow = head
        while fast.next:
            fast = fast.next.next
            slow = slow.next
            if not fast:
                return
            elif fast == slow:
                q = head
                while q != slow:
                    q    = q.next
                    slow = slow.next
                return q
        return

File: two point/142. Linked List Cycle II/solve_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode * fast = head;
        ListNode * slow = head;

        do
        {
            // 到顶
            if(!fast || !fast->next) return nullptr;
            fast = fast->next->next;
            slow = slow->next;
        }while(fast != slow);

        // 第一次相遇后，fast = head，然后再一步步相遇
        fast = head;
        while(fast != slow)
        {
            fast = fast->next;
            slow = slow->next;
        }

        return fast;
    }
};